x-pack/metricbeat/module/googlecloud/billing/billing.go

```diff
@@ -0,0 +1,273 @@
+		if err != nil {
+			return tables, err
+		}
+
+		// compare with given dataset_id
+		if dataset.DatasetID != datasetID {
+			continue
+		}
+
+		tit := dataset.Tables(ctx)
+		for {
+			var tableMeta tableMeta
+			table, err := tit.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				return tables, err
+			}
+
+			// make sure table ID fits the given table_pattern
+			if strings.HasPrefix(table.TableID, tablePattern) {
+				tableMeta.tableFullID = table.ProjectID + "." + table.DatasetID + "." + table.TableID
+				tableMeta.location = meta.Location
+				tables = append(tables, tableMeta)
+			}
+		}
+	}
+	return tables, nil
+}
+
+func (m *MetricSet) queryBigQuery(ctx context.Context, client *bigquery.Client, tableMeta tableMeta, month string, costType string) ([]mb.Event, error) {
+	var events []mb.Event
+	query := fmt.Sprintf(`
+			SELECT
+				invoice.month,
+				project.id,
+				cost_type,
+			  (SUM(CAST(cost * 1000000 AS int64))
+				+ SUM(IFNULL((SELECT SUM(CAST(c.amount * 1000000 as int64)) FROM UNNEST(credits) c), 0))) / 1000000
+				AS total_exact
+			FROM %s
+			WHERE project.id IS NOT NULL
+			AND invoice.month = '%s'
+			AND cost_type = '%s'
+			GROUP BY 1, 2, 3
+			ORDER BY 1 ASC, 2 ASC, 3 ASC;`, tableMeta.tableFullID, month, costType)
+
+	q := client.Query(query)
+	m.logger.Debug("bigquery query = ", query)
+
+	// Location must match that of the dataset(s) referenced in the query.
+	q.Location = tableMeta.location
+
+	// Run the query and print results when the query job is completed.
+	job, err := q.Run(ctx)
+	if err != nil {
+		err = fmt.Errorf("bigquery Run failed: %w", err)
+		m.logger.Error(err)
+		return events, err
+	}
+
+	status, err := job.Wait(ctx)
+	if err != nil {
+		err = fmt.Errorf("bigquery Wait failed: %w", err)
+		m.logger.Error(err)
+		return events, err
+	}
+
+	if err := status.Err(); err != nil {
+		err = fmt.Errorf("bigquery status error: %w", err)
+		m.logger.Error(err)
+		return events, err
+	}
+
+	it, err := job.Read(ctx)
+	for {
+		var row []bigquery.Value
+		err := it.Next(&row)
+		if err == iterator.Done {
+			break
+		}
+
+		if err != nil {
+			err = fmt.Errorf("bigquery RowIterator Next failed: %w", err)
+			m.logger.Error(err)
+			return events, err
+		}
+
+		if len(row) == 4 {
+			events = append(events, createEvents(row, m.config.ProjectID))
+		}
+	}
+	return events, nil
+}
+
+func createEvents(rowItems []bigquery.Value, accountID string) mb.Event {
+	event := mb.Event{}
+	event.MetricSetFields = common.MapStr{}
+	event.RootFields = common.MapStr{}
+
+	event.RootFields.Put("cloud.provider", "googlecloud")
+	event.RootFields.Put("cloud.account.id", accountID)
+	event.RootFields.Put("cloud.account.name", accountID)
+
+	event.MetricSetFields.Put("invoice_month", rowItems[0])
+	event.MetricSetFields.Put("project_id", rowItems[1])
+	event.MetricSetFields.Put("cost_type", rowItems[2])
+	event.MetricSetFields.Put("total", rowItems[3])
```

Comment:
nit: you could initialize all this together, no need to do a `Put` call right?

Reply:
haha not sure why I had to use `Put`... Will fix it in the next commit!! Thank you!!